specs/merkle-tree.md

```diff
@@ -122,16 +122,14 @@ To ensure that `APPROVE` and `CANCEL` leaves always happen first on a given chai
 ### 8. Each sequence of EXECUTE leaves with the same `chainId` must have the index of the first leaf be `1` and the index of the following leaves sequentially increment by `1`
 To ensure that all the `EXECUTE` leaves are executed in the correct order, they must have an `index` field that starts at `1` on each chain and sequentially increments by `1`.
 
-### 9. Merkle tree leaves must be ordered in the tree by the leaf's `index` and `chainId` fields ascending
-Sphinx Merkle tree leaves use an explicit `index` field in the leaf rather than an implicit index determined by the leaf's position in the tree. This is because the Sphinx Merkle tree is intended to be executed across multiple networks, and not all leaves will be executed on all networks. We impose the restriction that tree leaves must be ordered by their `index` and `chainId` fields ascending starting at the leftmost position within the tree.
+### 9. Merkle tree leaves must be ordered in the tree by the leaf's hash descending
+To allow for Sphinx Merkle trees to be reliably reproduced by off-chain tooling, we impose the restriction that Sphinx Merkle trees must have their leaves ordered within the tree by their hash descending.
 
-So, for example, if we intended to run a single transaction (one `EXECUTE` leaf) on Goerli (5), OP Goerli (420), and Polygon Mumbai (80001), then the leaves are expected to be ordered like so:
-1. `APPROVE` Leaf on Goerli (Chain ID 5)
-2. `EXECUTE` Leaf on Goerli
-3. `APPROVE` Leaf on OP Goerli (Chain ID 420)
-4. `EXECUTE` Leaf on OP Goerli
-5. `APPROVE` Leaf on Polyon Mumbai (Chain ID 80001)
-6. `EXECUTE` Leaf on Polygon Mumbai
+The leaf's hash is defined as the fields of the `SphinxLeaf` encoded as bytes and double-hashed using `keccak256`. For reference, the hash of the [SphinxLeaf](https://github.com/sphinx-labs/sphinx/blob/fdf4670089d026ad0bfd20e0ca2b46a02ee5a332/packages/contracts/contracts/core/SphinxDataTypes.sol#L28) can be calculated in Solidity like so:
+
+`keccak256(keccak256(abi.encode(leaf.chainId, leaf.index, leaf.leafType, leaf.data)))`.
```

Comment:
> `keccak256(keccak256(abi.encode(leaf.chainId, leaf.index, leaf.leafType, leaf.data)))`.

I'm not sure this is correct. Here's the hashing algorithm in `_getLeafHash`:

```
keccak256(abi.encodePacked(keccak256(abi.encode(_leaf))));
```

I don't know if the ABI encoded `SphinxLeaf` struct equals its ABI encoded fields. (I'm not familiar with the ABI encoding algorithm). Even if they are equal, it's not obvious that this is the case, so I think we should either change this code snippet or just link to the `_getLeafHash` function

Reply:
They are equal, but I agree it would be better to link to the `_getLeafHash` function.

Reply:
We don't ABI encode the fields when we call OpenZeppelin's library though. We use a tuple, which is a struct from ethers' POV:

```
  const tree = StandardMerkleTree.of(rawLeafArray, [
    'tuple(uint256, uint256, uint8, bytes)',
  ])
```

Reply:
https://docs.soliditylang.org/en/latest/abi-spec.html#encoding-of-indexed-event-parameters

Reply:
I wasn't able to find conclusive docs for or against this, but I tried out this alternative implementation which did not immediately work.

```
export const makeSphinxMerkleTreeFromLeaves = (
  leaves: SphinxLeaf[]
): SphinxMerkleTree => {
  const rawLeafArray = leaves.map((leaf) => Object.values(leaf))
  const tree = StandardMerkleTree.of(rawLeafArray, [
    'uint256',
    'uint256',
    'uint8',
    'bytes',
  ])

  return {
    root: tree.root,
    leavesWithProofs: leaves.map((leaf) => {
      const leafWithProof = {
        leaf,
        proof: tree.getProof(Object.values(leaf)),
      }
      return leafWithProof
    }),
  }
}
```

So to be safe, I'm rewording this to be very clear that we are ABI encoding a struct.